src/do_unmount/do_unmount.c

```diff
@@ -0,0 +1,382 @@
+    char *cmd;  /* Command with all arguments */
+    int length; /* Length of the command with all arguments */
+    int i;      /* Iterator */
+
+    /* Calculate the length of the command */
+    length = strlen(command) + 1; /* +1 for the null terminator */
+    for (i = 0; i < argc; i++)
+        length += strlen(argv[i]) + 3; /* +3 for the quotes and space */
+
+    /* Allocate memory for the command */
+    cmd = xmalloc(length * sizeof(char));
+    sprintf(cmd, "%s", command);
+
+    /* Craft the command with all passed arguments */
+    for (i = 0; i < argc; i++) {
+        strcat(cmd, " \"");
+        strcat(cmd, argv[i]);
+        strcat(cmd, "\"");
+    }
+
+    /* Open the command, free memory and return output file pointer for reading */
+    fp = popen(cmd, "r");
+    free(cmd);
+    return fp;
+}
+
+/* Populate the list of shared mounts in the system */
+static void
+populate_shared_list(RC_STRINGLIST **list)
+{
+    FILE *fp;               /* File pointer to the mountinfo file */
+    size_t len = 0;         /* Length of the line read */
+    char *line = NULL;      /* Line read from the mountinfo file */
+    char *token = NULL;     /* Token of the current line, no need to free this since is pointing to the original string */
+    char *path = NULL;      /* Path to relative to the current line */
+    int i;                  /* Fields iterator */
+
+    /* Initialize list */
+    *list = rc_stringlist_new();
+
+    /* Only linux supports shared mounts, so leave the list empty for eny other OS */
+    #ifdef __linux__
+    /* Open mountinfo file for reading */
+    fp = fopen("/proc/1/mountinfo", "r");
+    if (fp == NULL)
+        exit(EXIT_FAILURE);
+
+    /* Read the output a line at a time */
+    while (getline(&line, &len, fp) != -1) {
+        /* Split line by space */
+        token = strtok(line, " ");
+        i = 0;
+        while (token != NULL) {
+            if (i == 4)
+                xasprintf(&path, "%s", token);  /* Copy token into path */
+            else if (i == 6 && strstr(token, "shared:") != NULL) /* If token contains "shared:" */
+                rc_stringlist_add(*list, path);                  /* Note: if there is no optional field this token will be the separator character '-' */
+            token = strtok(NULL, " ");
+            i++;
+        }
+    }
+
+    /* Close file and free memory */
+    fclose(fp);
+    if (line)
+        free(line);
+    if (path)
+        free(path);
+    #endif
+}
+
+/* Pass arguments to mountinfo and store output in a list of paths to unmount */
+static int
+populate_unmount_list(RC_STRINGLIST **list, int argc, char **argv)
+{
+    int size = 0;       /* Number of paths to unmount */
+    FILE *fp;           /* File pointer to the output of the command */
+    char *path = NULL;  /* Path to add to the list */
+    size_t len = 0;     /* Length of the line read */
+
+
+    /* Open the command for reading */
+    fp = popen_vec("mountinfo", argc-2, argv+2);
```

Comment:
Shouldn't we just reuse the c code in `mountinfo.c` instead of calling an process and parsing it's output?

Reply:
This requires to edit `mountinfo.c` and create a new `mountinfo.h`. Is this what you are suggesting?

Reply:
Note that [in the original discussion](https://github.com/OpenRC/openrc/issues/662) it was explicitly requested a multi-thread program, separate from `mountinfo.c` that accesses `mountinfo.c` information.

Reply:
Now it is using code from libmountinfo.c